src/Security/Authentication/AuthenticationSuccessHandler.php

```diff
@@ -64,102 +67,68 @@ public function __construct(HttpUtils $httpUtils, ContaoFrameworkInterface $fram
-     *
-     * @param Request     $request
-     * @param BackendUser $user
+     * Triggers the postLogin hook.
      *
-     * @return RedirectResponse
+     * @param User $user
      */
-    private function handleBackendUser(Request $request, BackendUser $user): RedirectResponse
+    private function triggerPostLoginHook(User $user): void
     {
-        $this->triggerPostAuthenticateHook($user);
-
-        $route = $request->attributes->get('_route');
-
-        if ('contao_backend_login' !== $route) {
-            $parameters = [];
-
-            $routes = [
-                'contao_backend',
-                'contao_backend_preview',
-            ];
+        if (empty($GLOBALS['TL_HOOKS']['postLogin']) || !\is_array($GLOBALS['TL_HOOKS']['postLogin'])) {
+            return;
+        }
 
-            // Redirect to the last page visited upon login
-            if ($request->query->count() > 0 && \in_array($route, $routes, true)) {
-                $parameters['referer'] = base64_encode($request->getRequestUri());
-            }
+        @trigger_error('Using the "postLogin" hook has been deprecated and will no longer work in Contao 5.0. Use the security.interactive_login event instead.', E_USER_DEPRECATED);
 
-            return new RedirectResponse(
-                $this->router->generate('contao_backend_login', $parameters, UrlGeneratorInterface::ABSOLUTE_URL)
-            );
+        foreach ($GLOBALS['TL_HOOKS']['postLogin'] as $callback) {
+            $this->framework->createInstance($callback[0])->{$callback[1]}($user);
         }
-
-        return $this->httpUtils->createRedirectResponse($request, $this->determineTargetUrl($request));
     }
 
     /**
-     * Triggers the postAuthenticate hook.
+     * Stores the locale after user has logged in.
      *
-     * @param User $user
+     * @param Request $request
+     * @param string  $locale
      */
-    private function triggerPostAuthenticateHook(User $user): void
+    private function setLocale(Request $request, string $locale): void
     {
-        $this->framework->initialize();
-
-        if (empty($GLOBALS['TL_HOOKS']['postAuthenticate']) || !\is_array($GLOBALS['TL_HOOKS']['postAuthenticate'])) {
-            return;
+        if (null !== ($session = $request->getSession()) && $session->isStarted()) {
+            $session->set('_locale', $locale);
         }
 
-        @trigger_error('Using the postAuthenticate hook has been deprecated and will no longer work in Contao 5.0. Use the contao.post_authenticate event instead.', E_USER_DEPRECATED);
+        $request->setLocale($locale);
+        $this->translator->setLocale($locale);
 
-        foreach ($GLOBALS['TL_HOOKS']['postAuthenticate'] as $callback) {
-            $this->framework->createInstance($callback[0])->{$callback[1]}($user);
-        }
+        // Deprecated since Contao 4.0, to be removed in Contao 5.0
+        $GLOBALS['TL_LANGUAGE'] = str_replace('_', '-', $locale);
```

Comment:
Question: How is security related to locale handling here? Why did you have to move this?

Reply:
Because the `UserProvider` loads the user before the password is validated. Which caused the issue that the user language affected the locale even if the wrong password was entered (and e.g. changed the login screen language in the backend).